src/line_buffer.rs

```diff
@@ -194,13 +212,280 @@ impl LineBuffer {
+        }
+        let grapheme_1_start = self.grapheme_left_index();
+        let grapheme_2_end = self.grapheme_right_index();
+
+        if grapheme_1_start < insertion_offset && grapheme_2_end > insertion_offset {
+            let grapheme_1 = self.current_line()[grapheme_1_start..insertion_offset].to_string();
+            let grapheme_2 = self.current_line()[insertion_offset..grapheme_2_end].to_string();
+            self.replace_range(insertion_offset..grapheme_2_end, &grapheme_1);
+            self.replace_range(grapheme_1_start..insertion_offset, &grapheme_2);
+            self.insertion_point.offset = grapheme_2_end;
+        } else {
+            self.insertion_point.offset = insertion_offset;
+        }
+    }
 }
 
 /// Match any sequence of characters that are considered a word boundary
 fn is_word_boundary(s: &str) -> bool {
     !s.chars().any(char::is_alphanumeric)
 }
 
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    fn buffer_with(content: &str) -> LineBuffer {
+        let mut line_buffer = LineBuffer::new();
+        line_buffer.insert_str(content);
+
+        line_buffer
+    }
+
+    #[test]
+    fn test_new_buffer_is_empty() {
+        let line_buffer = LineBuffer::new();
+        assert!(line_buffer.is_empty())
+    }
+
+    #[test]
+    fn insert_str_updates_insertion_point_point_correctly() {
+        let mut line_buffer = LineBuffer::new();
+        line_buffer.insert_str("this is a command");
+
+        let expected_updated_insertion_point = InsertionPoint {
+            line: 0,
+            offset: 17,
+        };
+
+        assert_eq!(
+            expected_updated_insertion_point,
+            line_buffer.insertion_point()
+        );
+    }
+
+    #[test]
+    fn insert_char_updates_insertion_point_point_correctly() {
+        let mut line_buffer = LineBuffer::new();
+        line_buffer.insert_char('c');
+
+        let expected_updated_insertion_point = InsertionPoint { line: 0, offset: 1 };
+
+        assert_eq!(
+            expected_updated_insertion_point,
+            line_buffer.insertion_point()
+        );
+    }
+
+    #[test]
+    fn delete_left_grapheme_works() {
+        let mut line_buffer = buffer_with("This is a test");
+        line_buffer.delete_left_grapheme();
+
+        let expected_line_buffer = buffer_with("This is a tes");
+
+        assert_eq!(expected_line_buffer, line_buffer);
+    }
+
+    #[test]
+    fn delete_left_grapheme_works_with_emojis() {
+        let mut line_buffer = buffer_with("This is a test 😊");
+        line_buffer.delete_left_grapheme();
+
+        let expected_line_buffer = buffer_with("This is a test ");
+
+        assert_eq!(expected_line_buffer, line_buffer);
+    }
+
+    #[test]
+    fn delete_left_grapheme_on_an_empty_buffer_is_a_no_op() {
+        let mut line_buffer = buffer_with("");
+        line_buffer.delete_left_grapheme();
+
+        let expected_line_buffer = buffer_with("");
+
+        assert_eq!(expected_line_buffer, line_buffer);
```

Comment:
https://docs.rs/pretty_assertions/0.7.2/pretty_assertions/index.html can prettify the diffs. I think it would be a nice lib to add.

Reply:
I will try to limit the scope of this PR to this change only. Will add another PR that adds this dependency.